File: scripts/check_site.py
"""Check the static portfolio using only Python's standard library."""
import re
import sys
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import unquote, urlsplit


class Page(HTMLParser):
    def __init__(self, html):
        super().__init__()
        self.ids = []
        self.references = []
        self.headings = 0
        self.stack = []
        self.feed(html)
        assert not self.stack, f"Unclosed elements: {self.stack}"

    def handle_starttag(self, tag, attrs):
        if tag not in ("meta", "link", "img", "br", "hr", "input", "source", "wbr"):
            self.stack.append(tag)
        attrs = dict(attrs)
        if "id" in attrs:
            self.ids.append(attrs["id"])
        for key in ("src", "href"):
            if key in attrs:
                self.references.append(attrs[key])
        self.headings += tag == "h1"
        if tag == "img":
            assert attrs.get("alt"), "Images need descriptive alternative text"
        assert tag != "script", "This portfolio should work without JavaScript"

    def handle_endtag(self, tag):
        assert self.stack and self.stack.pop() == tag, f"Unbalanced element: {tag}"


if __name__ == "__main__":
    root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path(__file__).resolve().parents[1]
    html = (root / "index.html").read_text()
    css = (root / "styles.css").read_text()
    page = Page(html)
    assert page.headings == 1, "One main heading expected"
    assert len(page.ids) == len(set(page.ids)), "Duplicate section IDs"
    assert "<title>Farhan — Flutter Developer</title>" in html
    assert 'href="assets/farhan-mahi-cv.pdf" download' in " ".join(html.split())
    assert "mailto:sadiquefarhan038@gmail.com" in html
    assert not re.search(r"className|fetchPriority|_next/|@import|@theme", html + css)
    work = re.search(r'<section\b[^>]*id="work".*?</section>', html, re.S)
    assert work and "AI Limit Status" in work[0] and "One Ummah" not in work[0]
    references = page.references + re.findall(r'url\(["\']?([^"\')]+)', css)
    for reference in references:
        url = urlsplit(reference)
        if url.scheme or url.netloc:
            continue
        assert not url.path.startswith("/"), f"Use a relative path: {reference}"
        if url.path:
            assert (root / unquote(url.path)).is_file(), f"Missing file: {reference}"
        elif url.fragment:
            assert url.fragment in page.ids, f"Missing section: {reference}"
    print(f"Verified {len(references)} links/assets, metadata, image labels and personal-project scope.")

File: scripts/check-site.mjs
import assert from 'node:assert/strict';
import { existsSync, readFileSync } from 'node:fs';
import { resolve } from 'node:path';

const directory = resolve('dist/client');
const html = readFileSync(resolve(directory, 'index.html'), 'utf8');
assert.equal(
  (html.match(/<h1\b/g) ?? []).length,
  1,
  'One descriptive page heading',
);
assert.match(html, /Farhan — Flutter Developer/, 'Portfolio metadata');
assert.doesNotMatch(
  html,
  /Your site is taking shape|Untitled site|__next_error__/,
  'No starter or error page',
);
const ids = new Set(
  [...html.matchAll(/\bid="([^"]+)"/g)].map((match) => match[1]),
);
const references = [
  ...html.matchAll(/<(?:a|img|script|link)\b[^>]*\b(?:src|href)="([^"]+)"/g),
].map((match) => match[1]);
for (const reference of references) {
  if (reference.startsWith('#'))
    assert(ids.has(reference.slice(1)), `Anchor exists: ${reference}`);
  if (reference.startsWith('/') && !reference.startsWith('//')) {
    const pathname = decodeURIComponent(reference.split(/[?#]/)[0]);
    assert(
      existsSync(resolve(directory, '.' + pathname)),
      `Static asset exists: ${pathname}`,
    );
  }
}
const work = html.match(/<section\b[^>]*id="work"[\s\S]*?<\/section>/)?.[0];
assert(work, 'Selected work section exists');
assert.match(work, /AI Limit Status/, 'Personal project is featured');
assert.doesNotMatch(
  work,
  /One Ummah/,
  'Employer project stays out of personal work',
);
assert.match(
  html,
  /mailto:sadiquefarhan038@gmail.com/,
  'Working email contact',
);
assert.match(html, /farhan-mahi-cv.pdf/, 'CV download exists');
console.log(
  `Portfolio verified: metadata, personal-work scope, anchors and ${references.length} asset/link references.`,
);
